mining.go

```diff
@@ -958,6 +958,18 @@ func (g *BlkTmplGenerator) NewBlockTemplate(payToAddress dcrutil.Address) (*Bloc
 	stakeValidationHeight := g.chainParams.StakeValidationHeight
 
 	if nextBlockHeight >= stakeValidationHeight {
+		// Ensure the chain has enough live tickets to propagate after
+		// the next ticket maturity period from the current tip.
+		tickets := g.chain.LiveTicketsAtNextMaturity()
+		if tickets < uint64((g.chainParams.TicketsPerBlock/2)+1) {
+			futureHeight := best.Height + int64(g.chainParams.TicketMaturity) + 1
+			msg := fmt.Sprintf("chain continuity not guaranteed after "+
+				"height %d, remaining live tickets after ticket maturity "+
+				"period from current tip are: %d)", futureHeight, tickets)
+			minrLog.Debug(msg)
+			return nil, fmt.Errorf(msg)
+		}
+
```

Comment:
This seems only to be reactive rather than both proactive and reactive since it doesn't take the block template's ticket purchases into account. To not waste work on a dead-end chain is definitely part of the solution, but manual intervention to reorg the chain would still be required in a worst-case scenario. I think the warning signs for the ticket pool exhaustion should look around `TicketMaturity` blocks into the future and assume that all blocks would have 5 votes at least.

It seems like this problem should ultimately be solved by the consensus layer, and the objective of this PR is not clear as-is.

Reply:
Yes, the check ultimately should added at the consensus layer. However this implementation at the template generation level prevents to some extent those long reorgs that have occurred on testnet. You can think of this as a stop-gap solution until it's implemented at consensus level.